### PR TITLE
Fix some data and reader related API code

### PR DESCRIPTION
**Pull request by zhhsplendid**
Review and fix the example code in some reader related API doc. The fixed example codes were tested under Ubuntu Docker, v100 GPU, CUDA 9.0

Fix existing API example codes and screenshots:

paddle.fluid.io.PyReader

1:
ce22e2c9444cb251c2e4238f130c64bd

2:
bb978ee66ae2e605a8e12257653eac5a

3:
3f45d9c3ffcbc6905a5aa71a3f024bbc

paddle.fluid.layers.batch

c15b21df2aad90591e19df8965baef4b

paddle.fluid.layers.data

22742fba977c65ba2a41c60453f1af56

paddle.fluid.layers.Preprocessor

55bb6fbb1d8d8be2632ecfe8fb0d73ff

paddle.fluid.layers.py_reader

1. 
5e02fbafbd1f8e291471e1344b8a9915

2.
c43526d051036e1a85734b84a15edeac

3.
87b0dc6c01c90aa86cf50d171e72046b

paddle.fluid.program_guard

06b3dc5798419b9951f368d38da3a84c

paddle.layers.create_py_reader_by_data

77cf5f847374ab096e111e11279089e1

Add new example codes:

paddle.fluid.io.PyReader.decorate_batch_generator

aee8cf972d802f5e7b40589f492d18ce

paddle.fluid.io.PyReader.decorate_sample_generator

e6fef34e1c3f3170d1a7a5d01e759f61

paddle.fluid.io.PyReader.decorate_sample_list_generator

3f6a5eddf2ede6846f44346f59a5ff05

paddle.fluid.io.PyReader.reset

a75e731555707537a1fbfdcdc3768897

paddle.fluid.io.PyReader.start

459d1a87bf10cfe36709e38153977a25

test=develop

### CODE REVIEW

File: python/paddle/fluid/reader.py [hunk 1]
```diff
@@ -298,6 +381,36 @@ def decorate_sample_generator(self,
...
+                            yield np.random.uniform(low=0, high=255, size=[1, height, width])
+                    return generator
+
+                image = fluid.layers.data(name='image', shape=[784, 784], dtype='float32')
+                reader = fluid.io.PyReader(feed_list=[image], capacity=4, iterable=True)
+
+                user_defined_generator = random_image_generator(784, 784)
+                reader.decorate_sample_generator(user_defined_generator,
+                                                 batch_size=8,
+                                                 places=[fluid.core.CUDAPlace(0)])
```
Comment (sneaxiy) [inline python/paddle/fluid/reader.py]: `fluid.core.CUDAPlace(0)` -> `fluid.CUDAPlace(0)`.

File: python/paddle/fluid/reader.py [hunk 2]
```diff
@@ -298,6 +381,36 @@ def decorate_sample_generator(self,
...
+                # definition of network is omitted
+                executor = fluid.Executor(fluid.core.CUDAPlace(0))
```
Comment (sneaxiy) [inline python/paddle/fluid/reader.py]: `fluid.core.CUDAPlace(0)` -> `fluid.CUDAPlace(0)`.

File: python/paddle/fluid/reader.py [hunk 3]
```diff
@@ -364,6 +508,42 @@ def decorate_batch_generator(self, reader, places=None):
...
+
+                    return generator
+
+
+                image = fluid.layers.data(name='image', shape=[784, 784], dtype='float32')
+
+                reader = fluid.io.PyReader(feed_list=[image], capacity=4, iterable=True)
+
+                user_defined_generator = random_image_generator(784, 784)
+                reader.decorate_batch_generator(user_defined_generator,
```
Comment (sneaxiy) [inline python/paddle/fluid/reader.py]: When using `decorate_batch_generator`, reader should yield a tuple/list of numpy array, whose first element is with shape of `[batch_size, 784, 784]`. Please fix it.

File: python/paddle/fluid/reader.py [hunk 4]
```diff
@@ -364,6 +508,42 @@ def decorate_batch_generator(self, reader, places=None):
...
+                                                fluid.core.CUDAPlace(0))
```
Comment (sneaxiy) [inline python/paddle/fluid/reader.py]: Remove `core`.

File: python/paddle/fluid/reader.py [hunk 5]
```diff
@@ -364,6 +508,42 @@ def decorate_batch_generator(self, reader, places=None):
...
+                # definition of network is omitted
+                executor = fluid.Executor(fluid.core.CUDAPlace(0))
```
Comment (sneaxiy) [inline python/paddle/fluid/reader.py]: Remove `core`.

File: python/paddle/fluid/reader.py [hunk 6]
```diff
@@ -336,6 +449,37 @@ def decorate_sample_list_generator(self, reader, places=None):
...
+                            yield np.random.uniform(low=0,
+                                                    high=255,
+                                                    size=[1, 1, height, width])
+                    return generator
+
+                image = fluid.layers.data(name='image', shape=[784, 784], dtype='float32')
+                reader = fluid.io.PyReader(feed_list=[image], capacity=4, iterable=True)
+
+                user_defined_generator = random_image_generator(784, 784)
+                reader.decorate_sample_list_generator(user_defined_generator,
```
Comment (sneaxiy) [inline python/paddle/fluid/reader.py]: When using `decorate_sample_list_generator`, reader should return something like `([img1, ], [img2, ], ..., [imgn, 1]`. Please fix it.

File: python/paddle/fluid/reader.py (hunk 1 above)
Comment (sneaxiy) [reply python/paddle/fluid/reader.py]: When using `decorate_sample_generator`, reader should yield a list/tuple of sample data. Please fix it.

File: python/paddle/fluid/reader.py (hunk 5 above)
Comment (zhhsplendid) [reply python/paddle/fluid/reader.py]: Done

File: python/paddle/fluid/reader.py (hunk 4 above)
Comment (zhhsplendid) [reply python/paddle/fluid/reader.py]: Done

File: python/paddle/fluid/reader.py (hunk 3 above)
Comment (zhhsplendid) [reply python/paddle/fluid/reader.py]: Done

File: python/paddle/fluid/reader.py (hunk 2 above)
Comment (zhhsplendid) [reply python/paddle/fluid/reader.py]: Done

File: python/paddle/fluid/reader.py (hunk 1 above)
Comment (zhhsplendid) [reply python/paddle/fluid/reader.py]: Done

File: python/paddle/fluid/reader.py (hunk 6 above)
Comment (zhhsplendid) [reply python/paddle/fluid/reader.py]: Done

**Comment by zhhsplendid** [comment]
Please take another look

**Review by sneaxiy — APPROVED** [review]
LG API.

**Review by chengduoZH — APPROVED** [review]

**Review by shanyi15 — APPROVED** [review]
excellent job, very clear!

**Pull request merged**